src/main/java/carpet/utils/Messenger2.java

```diff
@@ -0,0 +1,245 @@
+            this.data.add(x);
+        }
+        this.size = this.data.size();
+        c = 0;
+    }
+
+
+    public Component getSomeComponents(){
+        return getSomeComponents("|");
+    }
+    public Component getSomeComponents(String str) {
+        TextComponent message = new TextComponent("");
+        while (true) {
+            if (isEnd(str)) {
+                return message;
+            }
+            Component com = getAComponent();
+            message.append(com);
+        }
+    }
+
+    private Component getAComponent() {
+        BaseComponent base = (BaseComponent) getARawComponent();
+        while (isModifier(now())) {
+            Object now = now();
+            base = (BaseComponent) getModifier(((String) now()).charAt(0)).apply(base, (String) now);
+        }
+        return base;
+    }
+
+    private BiFunction<BaseComponent, String, Component> getModifier(char ch) {
+        c += 1;
+        return (base, message) -> {
+            Style previousStyle = base.getStyle();
+            base.setStyle(switch (ch) {
+                case '?' -> previousStyle
+                        .withClickEvent(new ClickEvent(ClickEvent.Action.SUGGEST_COMMAND, message.substring(1)));
+                case '!' -> previousStyle
+                        .withClickEvent(new ClickEvent(ClickEvent.Action.RUN_COMMAND, message.substring(1)));
+                case '^' -> previousStyle
+                        .withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, message.equals("^")?getARawComponent():c(message.substring(1))));
+                case '@' -> previousStyle
+                        .withClickEvent(new ClickEvent(ClickEvent.Action.OPEN_URL, message.substring(1)));
+                case '&' -> previousStyle
+                        .withClickEvent(new ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, message.substring(1)));
+                default -> null;
+            });
+            return base;
+        };
+
+    }
+
+    private static boolean isModifier(Object peek) {
+        if (peek instanceof String pek) {
+            if (pek.startsWith("^") || pek.startsWith("?") || pek.startsWith("!") || pek.startsWith("@")
+                    || pek.startsWith("&")) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private Component getARawComponent() {
+        Component res;
+        if (now() instanceof Component b) {
+            res=  b;
+            c += 1;return res;
+        }
+        if (now() instanceof Entity ent) {
+            res=  (ent.getDisplayName());
+            c += 1;return res;
+        }
+        if (now() instanceof ItemStack ent) {
+            res= (ent.getDisplayName());
+            c += 1;return res;
+        }
+        if (now() instanceof String message) {
+            if (message.equalsIgnoreCase("")) {
+                c+=1;
+                return new TextComponent("");
+            }
+            if (message.equals("[")){
+                c+=1;
```

Comment:
to be honest, that is not necessary and kind of ugly, as u can simply put a Component as a arguments. 

but maybe some will enjoy in it as it is sometimes useful if u want to add click events to some components together.